src/game/client/gui/guicallbacks/controlbarcallback.cpp

```diff
@@ -12,18 +12,360 @@
+            pixel.y = ((data_1 >> 16) & 0xffff) - screen_y;
+            ICoord2D radar;
+
+            if ((!g_theRadar->Is_Radar_Hidden() || g_theRadar->Is_Radar_Forced())
+                && g_theRadar->Local_Pixel_To_Radar(&pixel, &radar)) {
+                const CommandButton *command = g_theInGameUI->Get_GUI_Command();
+
+                if (command != nullptr
+                    && (command->Get_Command() == GUI_COMMAND_SPECIAL_POWER
+                        || command->Get_Command() == GUI_COMMAND_SPECIAL_POWER_FROM_SHORTCUT)
+                    && (command->Get_Options() & COMMAND_OPTION_NEED_TARGET_POS) != 0) {
+                    MouseCursor index = g_theMouse->Get_Cursor_Index(command->Get_Cursor_Name());
+
+                    if (index == CURSOR_INVALID) {
+                        g_theMouse->Set_Cursor(CURSOR_TARGET);
+                    } else {
+                        g_theMouse->Set_Cursor(index);
+                    }
+                } else {
+                    const std::list<Drawable *> *drawables = g_theInGameUI->Get_All_Selected_Drawables();
+                    MouseCursor cursor = CURSOR_ARROW;
+
+                    if (!drawables->empty() && message != GWM_MOUSE_LEAVING) {
+                        if (command != nullptr && command->Get_Command() == GUI_COMMAND_ATTACK_MOVE) {
+                            cursor = CURSOR_ATTACK_MOVE;
+                        } else {
+                            cursor = CURSOR_MOVE;
+                        }
+                    }
+
+                    g_theMouse->Set_Cursor(cursor);
+                }
+            }
+
+            g_theInGameUI->Clear_Attack_Move_To_Mode();
+            return MSG_HANDLED;
+        }
+        default: {
+            return MSG_IGNORED;
+        }
+    }
+}
+
+WindowMsgHandledType Control_Bar_Input(GameWindow *window, unsigned int message, unsigned int data_1, unsigned int data_2)
+{
+    return MSG_IGNORED;
+}
+
+WindowMsgHandledType Control_Bar_System(GameWindow *window, unsigned int message, unsigned int data_1, unsigned int data_2)
+{
+    GameWindow *window2 = reinterpret_cast<GameWindow *>(data_1);
+    static NameKeyType buttonCommunicator;
+
+    if (g_theScriptEngine != nullptr && g_theScriptEngine->Is_End_Game_Timer_Running()) {
+        return MSG_IGNORED;
+    }
+
+    if (message > GBM_SELECTED_RIGHT) {
+        if (message == GEM_EDIT_DONE) {
+            int id = window2->Win_Get_Window_Id();
+            static const NameKeyType textID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:EditBeaconText");
+
+            if (id == textID && g_theInGameUI->Get_Select_Count() == 1) {
+                GameMessage *msg = g_theMessageStream->Append_Message(GameMessage::MSG_SET_BEACON_TEXT);
+                Utf16String text = Gadget_Text_Entry_Get_Text(window2);
+                g_theLanguageFilter->Filter_Line(text);
+                const unichar_t *str = text.Str();
+
+                while (str != nullptr && *str != '\0') {
```

Comment:
Condition looks odd. Probably should have been
```c++
if (str != nullptr)
  while (*str != '\0') {
    msg->Append_Wide_Char_Arg(*str);
    str++;
  }
```
Because str will not suddenly turn null.

Reply:
No idea why original does it this way.